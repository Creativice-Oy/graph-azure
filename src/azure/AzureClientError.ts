import { Response } from "node-fetch";

export default class AzureClientError extends Error {
  private static generateMessage(response: Response): string {
    const intro = "Unexpected response from Azure API:";
    const errorText = `'${response.status} - ${response.statusText}`;

    return [intro, errorText].join(" ");
  }

  public status: number;
  public statusText: string;

  constructor(response: Response) {
    super(AzureClientError.generateMessage(response));
    this.status = response.status;
    this.statusText = response.statusText;
  }
}
