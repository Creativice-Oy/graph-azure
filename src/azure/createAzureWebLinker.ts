import { AzureWebLinker } from "./types";

export default function createAzureWebLinker(
  defaultDomain: string | undefined,
): AzureWebLinker {
  return {
    portalResourceUrl: path => {
      if (defaultDomain && path) {
        return `https://portal.azure.com/#@${defaultDomain}/resource${path}`;
      }
    },
  };
}
