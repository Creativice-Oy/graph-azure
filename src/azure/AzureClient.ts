import querystring from "querystring";

import { IntegrationLogger } from "@jupiterone/jupiter-managed-integration-sdk";

import AzureClientError from "./AzureClientError";
import authenticate from "./graph/authenticate";
import { Group, GroupMember, User } from "./types";

export enum Method {
  GET = "get",
  POST = "post",
}

export interface PaginationOptions {
  limit?: number;
  nextLink?: string;

  /**
   * The property names for `$select` query param.
   */
  select?: string | string[];
}

interface ResourceUrlOptions extends PaginationOptions {
  path: string;
}

interface FetchResourcesInput {
  url: string;
}

export interface FetchResourcesResponse<T extends microsoftgraph.Entity> {
  nextLink: string | undefined;
  resources: T[];
  err?: AzureClientError;
}

export default class AzureClient {
  private clientId: string;
  private clientSecret: string;
  private directoryId: string;
  private accessToken: string;
  private apiHost = "https://graph.microsoft.com";
  private apiVersion = "v1.0";
  private logger: IntegrationLogger;

  constructor(
    clientId: string,
    clientSecret: string,
    directoryId: string,
    logger: IntegrationLogger,
  ) {
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.directoryId = directoryId;
    this.logger = logger;
  }

  public async authenticate() {
    this.accessToken = await authenticate({
      clientId: this.clientId,
      clientSecret: this.clientSecret,
      directoryId: this.directoryId,
    });
  }

  public async fetchUsers(
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<User>> {
    return this.fetchResources({
      url: this.resourceUrl({ path: "/users", ...options }),
    });
  }

  public async fetchGroups(
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<Group>> {
    return this.fetchResources({
      url: this.resourceUrl({ path: "/groups", ...options }),
    });
  }

  public async fetchGroupMembers(
    groupId: string,
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<GroupMember>> {
    return this.fetchResources({
      url: this.resourceUrl({
        path: `/groups/${groupId}/members`,
        ...options,
      }),
    });
  }

  private async fetchResources<T>(
    input: FetchResourcesInput,
  ): Promise<FetchResourcesResponse<T>> {
    try {
      this.logger.trace({ url: input.url }, "Fetching Azure resources");

      const response = await this.makeRequest<any>(input.url, Method.GET);

      return {
        nextLink: response["@odata.nextLink"],
        resources: response.value,
      };
    } catch (err) {
      if (err.status === 404) {
        return { nextLink: undefined, resources: [] };
      } else {
        this.logger.warn(
          { err, url: input.url, method: Method.GET },
          "Azure resource request failed",
        );

        return { nextLink: undefined, resources: [], err };
      }
    }
  }

  /**
   * Pagination: https://docs.microsoft.com/en-us/graph/paging
   * Throttling with retry after: https://docs.microsoft.com/en-us/graph/throttling
   * Batching requests: https://docs.microsoft.com/en-us/graph/json-batching
   */
  private async makeRequest<T>(
    url: string,
    method: Method,
    headers?: {},
  ): Promise<T> {
    if (!this.accessToken) {
      await this.authenticate();
    }

    const options: RequestInit = {
      method,
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
        Authorization: `Bearer ${this.accessToken}`,
        ...headers,
      },
    };

    const response = await fetch(url, options);
    if (response.status >= 300) {
      throw new AzureClientError(response);
    }

    return response.json();
  }

  /**
   * Constructs a resource url to include optional query params.
   *
   * `fetch` will itself perform url escaping, so this function must not do so
   * itself. For example, the `url` library has a nice API for constructing
   * parameters, but the `url.toString()` function answers excaped strings,
   * which `fetch` then attempts to escape, making the params invalid.
   */
  private resourceUrl(options: ResourceUrlOptions): string {
    if (options.nextLink) {
      return options.nextLink;
    } else {
      const queryParams: { [key: string]: any } = {};

      if (options.limit) {
        queryParams.$top = options.limit;
      }

      if (options.select) {
        queryParams.$select = Array.isArray(options.select)
          ? options.select.join(",")
          : options.select;
      }

      let url = `${this.apiHost}/${this.apiVersion}${options.path}`;
      if (Object.keys(queryParams).length > 0) {
        url +=
          "?" +
          querystring.stringify(queryParams, undefined, undefined, {
            encodeURIComponent: (s: string) => s,
          });
      }

      return url;
    }
  }
}
