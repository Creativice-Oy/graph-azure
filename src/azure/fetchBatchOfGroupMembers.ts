import {
  IntegrationCacheEntry,
  IntegrationStepIterationState,
} from "@jupiterone/jupiter-managed-integration-sdk";
import { Group } from "@microsoft/microsoft-graph-types";

import { AzureExecutionContext, GroupsCacheState } from "../types";
import { getBatchPages, getPageLimit } from "./batch";
import { GroupMember } from "./types";
import AzureClientError from "./AzureClientError";

export default async function fetchBatchOfGroupMembers(
  executionContext: AzureExecutionContext,
  iterationState: IntegrationStepIterationState,
): Promise<IntegrationStepIterationState> {
  const { azure, logger } = executionContext;
  const cache = executionContext.clients.getCache();
  const groupsCache = cache.iterableCache<
    IntegrationCacheEntry,
    GroupsCacheState
  >("groups");

  const batchPages = getBatchPages("group_members", 10);
  const limit = getPageLimit("group_members", 200);

  let membersCount: number = iterationState.state.count || 0;
  let nextLink: string | undefined = iterationState.state.nextLink;
  let groupIndex: number = iterationState.state.groupIndex || 0;

  let pagesProcessed = 0;
  let totalGroups = 0;
  let fetchError: AzureClientError | undefined;

  logger.info(
    { totalGroups, groupIndex },
    "Starting group iteration for members fetch",
  );

  await groupsCache.forEach(
    async (e) => {
      const groupEntry = e.entry;
      const groupEntryIndex = e.entryIndex;
      const group = groupEntry.data as Group;

      totalGroups = e.totalEntries;

      // load up accumulated list of members
      const groupMembers: GroupMember[] =
        (group.members as GroupMember[]) || [];

      logger.info(
        { totalGroups, groupEntryIndex, groupIndex },
        "Starting group members fetch loop",
      );

      // fetch members for group
      do {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const response = await azure.fetchGroupMembers(group.id!, {
          nextLink,
          limit,
          // Changes here should be reflected in `GroupMember`
          select: ["id", "displayName", "jobTitle", "mail"],
        });

        if (!response.err) {
          groupMembers.push(...response.resources);
          membersCount += response.resources.length;
          nextLink = response.nextLink;
        } else {
          fetchError = response.err;
        }

        pagesProcessed++;
      } while (!fetchError && pagesProcessed < batchPages && nextLink);

      logger.info(
        { fetchError, nextLink, pagesProcessed },
        "Completed group members fetch loop",
      );

      if (fetchError) {
        // Stop iterating groups altogether until we learn more about the errors.
        // Are there some groups we don't have access to, so that we should skip
        // them, or is this a temporary problem that should be ignored (do not
        // produce any operations to avoid unintended data deletion)?
        return true;
      } else {
        // update the cached group to include members loaded during this invocation
        await groupsCache.putEntries([
          {
            key: groupEntry.key,
            data: { ...groupEntry.data, members: groupMembers },
          },
        ]);

        // move to the next group if there are no more members in this group
        if (!nextLink) {
          groupIndex = groupEntryIndex + 1;
        }

        // stop iteration when pagesProcessed reaches invocation limit
        return pagesProcessed >= batchPages;
      }
    },
    { skip: groupIndex },
  );

  const groupMembersFetchCompleted =
    typeof nextLink !== "string" && groupIndex >= totalGroups;

  const finished = !!fetchError || groupMembersFetchCompleted;

  logger.info(
    {
      fetchError,
      nextLink,
      pagesProcessed,
      totalGroups,
      groupIndex,
      groupMembersFetchCompleted,
      finished,
    },
    "Completed group iteration for members fetch",
  );

  await groupsCache.putState({
    groupMembersFetchCompleted: !fetchError && groupMembersFetchCompleted,
    fetchError: fetchError,
  });

  return {
    ...iterationState,
    finished,
    state: {
      nextLink,
      limit,
      pages: pagesProcessed,
      count: membersCount,
      groupIndex,
    },
  };
}
