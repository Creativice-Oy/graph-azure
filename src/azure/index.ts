import createAzureWebLinker from "./createAzureWebLinker";

export { createAzureWebLinker };

export * from "./types";
