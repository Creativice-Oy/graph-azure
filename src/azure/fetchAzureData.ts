import AzureClient from "./AzureClient";
import { AzureDataModel, Group, GroupMembers } from "./types";

export default async function fetchAzureData(
  client: AzureClient,
): Promise<AzureDataModel> {
  const [groups, users] = await Promise.all([
    client.fetchGroups(),
    client.fetchUsers(),
  ]);

  const groupsMembers = await Promise.all(
    groups.map((group: Group) => fetchGroupMembers(group, client)),
  );

  return { groups, groupsMembers, users };
}

async function fetchGroupMembers(
  group: Group,
  client: AzureClient,
): Promise<GroupMembers> {
  const members = await client.fetchMembers(group.id);

  return {
    group,
    members,
  };
}
