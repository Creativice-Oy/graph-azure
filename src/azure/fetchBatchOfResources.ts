import {
  IntegrationCacheEntry,
  IntegrationStepExecutionResult,
  IntegrationStepIterationState,
} from "@jupiterone/jupiter-managed-integration-sdk";

import { AzureExecutionContext, ResourceCacheState } from "../types";
import AzureClient, {
  FetchResourcesResponse,
  PaginationOptions,
} from "./AzureClient";
import { getBatchPages, getPageLimit } from "./batch";

/**
 * An iterating execution handler that loads resources and associated data in
 * batches of pages of resources, storing the raw response data in the
 * `IntegrationCache` for later processing in another step.
 */
export default async function fetchBatchOfResources<
  T extends microsoftgraph.Entity
>(
  executionContext: AzureExecutionContext,
  iterationState: IntegrationStepIterationState,
  resourceName: string,
  fetch: (
    azure: AzureClient,
    paginationOptions: PaginationOptions,
  ) => Promise<FetchResourcesResponse<T> | undefined>,
): Promise<IntegrationStepExecutionResult> {
  const { azure } = executionContext;
  const cache = executionContext.clients.getCache();
  const resourceCache = cache.iterableCache<
    IntegrationCacheEntry,
    ResourceCacheState
  >(resourceName);

  const batchPages = getBatchPages(resourceName, 2);
  const limit = getPageLimit(resourceName, 200);

  let pagesProcessed = 0;
  let entryCount: number = iterationState.state.count || 0;
  let nextLink: string | undefined = iterationState.state.nextLink;

  do {
    const response = await fetch(azure, {
      nextLink,
      limit,
    });

    if (response) {
      entryCount = await resourceCache.putEntries(
        response.resources.map(e => ({
          key: e.id!,
          data: e,
        })),
      );
      nextLink = response.nextLink;
    }

    pagesProcessed++;
  } while (pagesProcessed < batchPages && nextLink);

  const finished = typeof nextLink !== "string";
  await resourceCache.putState({ resourceFetchCompleted: finished });

  return {
    iterationState: {
      ...iterationState,
      finished,
      state: {
        nextLink,
        limit,
        pages: pagesProcessed,
        count: entryCount,
      },
    },
  };
}
