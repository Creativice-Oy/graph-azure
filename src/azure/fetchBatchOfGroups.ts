import {
  IntegrationStepExecutionResult,
  IntegrationStepIterationState,
} from "@jupiterone/jupiter-managed-integration-sdk";

import { AzureExecutionContext } from "../types";
import fetchBatchOfResources from "./fetchBatchOfResources";

export default async function fetchBatchOfGroups(
  executionContext: AzureExecutionContext,
  iterationState: IntegrationStepIterationState,
): Promise<IntegrationStepExecutionResult> {
  return fetchBatchOfResources(
    executionContext,
    iterationState,
    "groups",
    (azure, options) => azure.fetchGroups(options),
  );
}
