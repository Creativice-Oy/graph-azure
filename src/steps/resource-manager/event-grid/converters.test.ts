import { createAzureWebLinker } from '../../../azure';
import {
  createEventGridDomainEntity,
  createEventGridDomainTopicEntity,
  createEventGridTopicSubscriptionEntity,
  createEventGridTopicEntity,
} from './converters';
import {
  Domain,
  DomainTopic,
  EventSubscription,
  Topic,
} from '@azure/arm-eventgrid/esm/models';

const webLinker = createAzureWebLinker('something.onmicrosoft.com');

describe('createEventGridDomainEntity', () => {
  test('properties transferred', () => {
    const data: Domain = {
      endpoint:
        'https://j1dev-event-grid-domain.eastus-1.eventgrid.azure.net/api/events',
      id:
        '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resoruce/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain',
      inputSchema: 'EventGridSchema',
      location: 'eastus',
      metricResourceId: '10000000-0000-0000-0000-000000000000',
      name: 'j1dev-event-grid-domain',
      provisioningState: 'Succeeded',
      publicNetworkAccess: 'Enabled',
      tags: {
        environment: 'j1dev',
      },
      type: 'Microsoft.EventGrid/domains',
    };

    const domainEntity = createEventGridDomainEntity(webLinker, data);

    expect(domainEntity).toMatchSnapshot();
    expect(domainEntity).toMatchGraphObjectSchema({
      _class: ['Service'],
      schema: {},
    });
  });
});

describe('createEventGridDomainTopicEntity', () => {
  test('properties transferred', () => {
    const data: DomainTopic = {
      id:
        '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain/topics/j1dev-event-grid-domain-topic',
      name: 'j1dev-event-grid-domain-topic',
      provisioningState: 'Succeeded',
      type: 'Microsoft.EventGrid/domains/topics',
    };

    const domainTopicEntity = createEventGridDomainTopicEntity(webLinker, data);

    expect(domainTopicEntity).toMatchSnapshot();
    expect(domainTopicEntity).toMatchGraphObjectSchema({
      _class: ['Queue'],
      schema: {},
    });
  });
});

describe('createEventGridTopicSubscriptionEntity', () => {
  test('properties transferred', () => {
    const data: EventSubscription = {
      id:
        '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic/providers/Microsoft.EventGrid/eventSubscriptions/j1dev-event-grid-subscription',
      name: 'j1dev-event-grid-subscription',
      provisioningState: 'Succeeded',
      type: 'Microsoft.EventGrid/eventSubscriptions',
    };

    const eventSubscriptionEntity = createEventGridTopicSubscriptionEntity(
      webLinker,
      data,
    );

    expect(eventSubscriptionEntity).toMatchSnapshot();
    expect(eventSubscriptionEntity).toMatchGraphObjectSchema({
      _class: ['Subscription'],
      schema: {},
    });
  });
});

describe('createEventGridTopicEntity', () => {
  test('properties transferred', () => {
    const data: Topic = {
      endpoint:
        'https://j1dev-eventgrid-topic.eastus-1.eventgrid.azure.net/api/events',
      id:
        '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic',
      inputSchema: 'EventGridSchema',
      location: 'eastus',
      metricResourceId: '20000000-0000-0000-0000-000000000000',
      name: 'j1dev-event-grid-topic',
      provisioningState: 'Succeeded',
      publicNetworkAccess: 'Enabled',
      tags: { environment: 'j1dev' },
      type: 'Microsoft.EventGrid/topics',
    };

    const topicEntity = createEventGridTopicEntity(webLinker, data);

    expect(topicEntity).toMatchSnapshot();
    expect(topicEntity).toMatchGraphObjectSchema({
      _class: ['Queue'],
      schema: {},
    });
  });
});
