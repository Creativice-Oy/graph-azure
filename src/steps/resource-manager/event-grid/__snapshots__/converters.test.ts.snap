// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEventGridDomainEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Service",
  ],
  "_key": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resoruce/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "endpoint": "https://j1dev-event-grid-domain.eastus-1.eventgrid.azure.net/api/events",
        "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resoruce/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain",
        "inputSchema": "EventGridSchema",
        "location": "eastus",
        "metricResourceId": "10000000-0000-0000-0000-000000000000",
        "name": "j1dev-event-grid-domain",
        "provisioningState": "Succeeded",
        "publicNetworkAccess": "Enabled",
        "tags": Object {
          "environment": "j1dev",
        },
        "type": "Microsoft.EventGrid/domains",
      },
    },
  ],
  "_type": "azure_event_grid_domain",
  "category": Array [
    "infrastructure",
  ],
  "createdOn": undefined,
  "displayName": "j1dev-event-grid-domain",
  "endpoint": "https://j1dev-event-grid-domain.eastus-1.eventgrid.azure.net/api/events",
  "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resoruce/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain",
  "inputSchema": "EventGridSchema",
  "location": "eastus",
  "metricResourceId": "10000000-0000-0000-0000-000000000000",
  "name": "j1dev-event-grid-domain",
  "provisioningState": "Succeeded",
  "publicNetworkAccess": "Enabled",
  "tag.environment": "j1dev",
  "type": "Microsoft.EventGrid/domains",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resoruce/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain",
}
`;

exports[`createEventGridDomainTopicEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Queue",
  ],
  "_key": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain/topics/j1dev-event-grid-domain-topic",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain/topics/j1dev-event-grid-domain-topic",
        "name": "j1dev-event-grid-domain-topic",
        "provisioningState": "Succeeded",
        "type": "Microsoft.EventGrid/domains/topics",
      },
    },
  ],
  "_type": "azure_event_grid_domain_topic",
  "createdOn": undefined,
  "displayName": "j1dev-event-grid-domain-topic",
  "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain/topics/j1dev-event-grid-domain-topic",
  "name": "j1dev-event-grid-domain-topic",
  "provisioningState": "Succeeded",
  "type": "Microsoft.EventGrid/domains/topics",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/domains/j1dev-event-grid-domain/topics/j1dev-event-grid-domain-topic",
}
`;

exports[`createEventGridTopicSubscriptionEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Subscription",
  ],
  "_key": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic/providers/Microsoft.EventGrid/eventSubscriptions/j1dev-event-grid-subscription",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic/providers/Microsoft.EventGrid/eventSubscriptions/j1dev-event-grid-subscription",
        "name": "j1dev-event-grid-subscription",
        "provisioningState": "Succeeded",
        "type": "Microsoft.EventGrid/eventSubscriptions",
      },
    },
  ],
  "_type": "azure_event_grid_event_subscription",
  "createdOn": undefined,
  "displayName": "j1dev-event-grid-subscription",
  "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic/providers/Microsoft.EventGrid/eventSubscriptions/j1dev-event-grid-subscription",
  "name": "j1dev-event-grid-subscription",
  "provisioningState": "Succeeded",
  "type": "Microsoft.EventGrid/eventSubscriptions",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic/providers/Microsoft.EventGrid/eventSubscriptions/j1dev-event-grid-subscription",
}
`;

exports[`createEventGridTopicEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Queue",
  ],
  "_key": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "endpoint": "https://j1dev-eventgrid-topic.eastus-1.eventgrid.azure.net/api/events",
        "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic",
        "inputSchema": "EventGridSchema",
        "location": "eastus",
        "metricResourceId": "20000000-0000-0000-0000-000000000000",
        "name": "j1dev-event-grid-topic",
        "provisioningState": "Succeeded",
        "publicNetworkAccess": "Enabled",
        "tags": Object {
          "environment": "j1dev",
        },
        "type": "Microsoft.EventGrid/topics",
      },
    },
  ],
  "_type": "azure_event_grid_topic",
  "createdOn": undefined,
  "displayName": "j1dev-event-grid-topic",
  "endpoint": "https://j1dev-eventgrid-topic.eastus-1.eventgrid.azure.net/api/events",
  "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic",
  "inputSchema": "EventGridSchema",
  "location": "eastus",
  "metricResourceId": "20000000-0000-0000-0000-000000000000",
  "name": "j1dev-event-grid-topic",
  "provisioningState": "Succeeded",
  "publicNetworkAccess": "Enabled",
  "tag.environment": "j1dev",
  "type": "Microsoft.EventGrid/topics",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/j1dev/providers/Microsoft.EventGrid/topics/j1dev-event-grid-topic",
}
`;
